Detect insufficient resources and deadlocks more usefully

adamnovak opened the pull request:
This fixes #3016. 
That issue is really caused by Cactus setting the deadlock checking time too high, so it takes us an hour to even see the oops-all-services deadlock, and another hour to do anything about it.

This PR changes Toil to separate the time we wait to terminate the workflow from the time we wait to see if there's a deadlock. It adds some warning-level log messages when the deadlock checker *starts* to think there's a deadlock, and it is able to pull a hint from the batch system (currently only implemented for `SingleMachineBatchSystem`) that can suggest to the user why the deadlock might be happening.

With Cactus installed, I did this to produce a service deadlock:

```
mkdir -p ~/mnt/tmpfs
sudo mount -t tmpfs -o size=2200000000 tmpfs ~/mnt/tmpfs
TMPDIR=$HOME/mnt/tmpfs cactus --workDir=$HOME/mnt/tmpfs ./tree examples/evolverMammals.txt examples/evolverMammals.hal --root mr
```

And now I get these messages about it:

```
Issued job 'CactusSetupPhase' kind-CactusSetupPhase/instance-juylrkhr with job batch system ID: 2 and cores: 1, disk: 2.0 G, and memory: 3.3 G
Potential deadlock detected! All 1 running jobs are service jobs, with no normal jobs to use them! The batch system reports: Not enough disk to run job 2
Potentially deadlocked for 30 seconds. Waiting at most 3570 more seconds for any of 1 issued non-service jobs to schedule and start. The batch system reports: Not enough disk to run job 2
Potentially deadlocked for 60 seconds. Waiting at most 3540 more seconds for any of 1 issued non-service jobs to schedule and start. The batch system reports: Not enough disk to run job 2
...
```

It might be noisy in big cluster runs, but you can (manually) up `--deadlockCheckInterval` if you know you're on a batch system that takes a while to schedule jobs and don't want to hear about it. It might make sense to up it (and the deadlock wait too) when running with autoscaling, automatically, if we can work out how.

adamnovak commented:
@diekhans Does this inspire you to bliss?

diekhans commented:
It would inspire me to wait and see what happens :-)

Adam Novak <notifications@github.com> writes:
> @diekhans Does this inspire you to bliss?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/DataBiosphere/toil/pull/3043#issuecomment-618006631
> @diekhans Does this inspire you to bliss?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.*
>

DailyDreaming commented on src/toil/batchSystems/__init__.py:
```diff
@@ -12,19 +12,32 @@
-from builtins import object
 import sys
 
-if sys.version_info >= (3, 0):
```
It always feels good seeing these removed.  Thanks!

DailyDreaming commented on src/toil/batchSystems/__init__.py:
```diff
@@ -12,19 +12,32 @@
+        return self.msg
+
+@total_ordering
+class MemoryString:
```
Created an issue based on this since I think some of this is duplicated in a few places: https://github.com/DataBiosphere/toil/issues/3046

DailyDreaming commented on src/toil/batchSystems/abstractBatchSystem.py:
```diff
@@ -180,6 +180,26 @@ def getUpdatedBatchJob(self, maxWait):
+        
+    def getSchedulingHint(self):
+        """
+        Get a hint for the user about anything that might be going wrong in the
```
Hints in CWL are resource requirements, so to avoid confusion, could this explicitly state somewhere in the first sentence that this is intended to be a log message?  Something like:

s/`Get a hint for the user`/`Get a log message string serving as a hint for the user`/g ?

adamnovak commented:
OK, I've revised this in light of your suggestions @DailyDreaming. I've changed the documentation around for the deadlock checking time parameters, so it should be clearer what they are actually there for, and I've replaced the "hints" with "details" and "status messages".
